### PR TITLE
Update to sqlite3 5.1.6

### PR DESCRIPTION
**Pull request by benallfree**
This PR brings @types/sqlite3 up to date with the latest version of the sqlite3 package.

sqlite3 ships with its own type definitions https://github.com/TryGhost/node-sqlite3/blob/master/lib/sqlite3.d.ts, so this is just a copy of that.

This @types/sqlite3 package remains important because the sqlite3 interface has become a defacto-standard and is used by projedts like node-sqlite. @types/sqlite3 makes it possible to reference the sqlite3 interface typings without pulling in the entirety of the sqlite3 package as a dependency. See https://github.com/kriasoft/node-sqlite/pull/177 and https://github.com/kriasoft/node-sqlite/issues/175 for further discussion.

I would also be interested in becoming an owner of @types/sqlite3 so I can keep it up to date with the sqlite3 project and even extend it with missing typings such as https://github.com/TryGhost/node-sqlite3/pull/1726.

### CODE REVIEW

**Comment by benallfree** [comment]
The failing tests are not part of this package.

File: types/sqlite3/package.json
Comment (theogravity) [inline types/sqlite3/package.json:6]: nit: https://github.com/TryGhost/node-sqlite3

**Review by theogravity — APPROVED** [review]

**Comment by theogravity** [comment]
@benallfree can you check that the spatialite test failure doesn't have a dependency on the types from this package?

**Review by theogravity — CHANGES_REQUESTED** [review]
Pending approval waiting for feedback on if the failing test is due to a dependency issue.

**Comment by benallfree** [comment]
Yes everything passes locally but I'll check again.

**Comment by theogravity** [comment]
> Yes everything passes locally but I'll check again.

Try running the test in that affected package

**Comment by benallfree** [comment]
Yes, `spatialite` depends on `sqlite3` and fails. Its tests need the same typing updates as this package.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/spatialite/spatialite-tests.ts
1. types/sqlite3/sqlite3-tests.ts
1. types/sqlite3/index.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d6f9470459e21fbdfc2ed02bdf08b66a8af0174a
</p>

**Comment by theogravity** [comment]
@benallfree fix the lint issues and this should be good to go thanks

**Review by theogravity — APPROVED** [review]

**Comment by benallfree** [comment]
@atd-schubert @nmalaguti @dpyro Any thoughts on this PR? Also I'd be happy to take an owner role if possible.

**Review by gabritto — COMMENTED** [review]
The changes look good, but it's unfortunate to have the type definitions duplicated on DT and on the original package, as it generates work to keep them up to date. Ideally we'd take it out of DT, but I understand this might not be ideal for everyone.

I left a minor comment, and also, if you want, you can add yourself as an owner in the `package.json` before I approve this PR.

File: types/sqlite3/index.d.ts
Comment (benallfree) [reply types/sqlite3/index.d.ts]: Done 👍

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/sqlite3/package.json

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 784119d097bc1d01c17e77e134bf99530fc2333a
</p>

**Comment by benallfree** [comment]
@gabritto Added owner info and removed `.d.ts` header - thank you!

**Review by theogravity — APPROVED** [review]

**Comment by jakebailey** [comment]
> This @types/sqlite3 package remains important because the sqlite3 interface has become a defacto-standard and is used by projedts like node-sqlite. @types/sqlite3 makes it possible to reference the sqlite3 interface typings without pulling in the entirety of the sqlite3 package as a dependency.

Truthfully, I'm not sure that this is a good idea; if someone installs the real sqlite3 package, then that package's types will take precedence and the DT types won't work. So if the two are ever out of date, it will be misleading to anyone who has managed to only install the package from DT. Not having the DT package at all is a lot easier to reason about.

I'm also not sure what the point is of trying to make things that optimal; surely _someone_ is going to have to install `sqlite3` at some point, because otherwise why use the types at all? DT describes actual runtime values and shouldn't be used as grab bag of random types.

**Comment by theogravity** [comment]
I'm the maintainer of the sqlite package and I do not include but copy the sqlite3 types because I don't want to make sqlite3 a direct dependency of my project (so people can use the lib with a compatible API interface) but still need the types from it.

Having the types come from here allows me to use it as a dep without requiring the actual sqlite3 binaries and allows other developers who wish to extend the types to add their own functionality as the current way I do it (copy the type def file into my project) prevents that from working.

**Comment by andrewbranch** [comment]
I think it’s a pretty bad idea to duplicate the types between the runtime package and DT. Instead, you might consider publishing a separate `sqlite3-types` or something.

**Comment by benallfree** [comment]
@andrewbranch Do you think this package should be deprecated then?

**Comment by andrewbranch** [comment]
Things will not get any _worse_ by merging this PR, so I will approve it if everyone is intent on going that route. But I read all the linked discussions and am still convinced that this approach will be more harmful than helpful. It seems like publishing a separate package outside of DT would address @theogravity’s goals without duplicating code or being tangled up in TypeScript’s special module resolution behavior for `@types` packages, so I think that should be considered.

**Comment by benallfree** [comment]
@andrewbranch That's an interesting perspective, thank you. I've started a discussion on https://github.com/TryGhost/node-sqlite3/issues/1732 to see if they have any interest in reviving @types/sqlite3 as the official source for typings.

**Comment by jakebailey** [comment]
I'm confused by the above; my reading is that Andrew was saying to _not_ do that, and instead create an npm package not on DT that would contain the types, not keep this package as the official source.

**Comment by benallfree** [comment]
Only because sqlite3 already has the types in their repo. I was suggesting that if sqlite3 wanted to move their types back to here, then this could become the official package again rather than having a separate package.

At this point, it doesn't seem like anyone wants to move away from their approaches so perhaps deprecating this package is the best way to close this issue. (Although it won't solve the original problem)

**Comment by benine203** [comment]
there has got to be plans for making a build mode called 'upstream datatypes puller'
it gets the tagged version, append the .9999 and everything

**Comment by benallfree** [comment]
@benine203 Are you proposing that @types/sqlite3 be automatically updated whenever the sqlite3 types change?

**Comment by benine203** [comment]
yeah indeed so but doing may need some non-existent build infrastructure
capabilities. to kick a new upstream-based release in motion:

a committer just initially converts a typing module from regular to
upstream-based by supplying a set of patches or middleware to
extract/mangle the upstream typedefs which don't change often but what
changes on each PR is either a URL/sha256 pair like in BSD's pkgsrc or a
remote/tag pair. the CI has added machinery to interpret this declarative
update and add actual commits to the PR and after that it is like before
and maintainers can fix/tweak before final acceptance.

or it can be all updated on some periodic puller on a long-lived instance
somewhere that ETLs the upstream source and files PRs. if issues in PRs
arise sends out notifications to a list of module owners to come review.

testing should be less involved as the upstream is supposed to have done
this already.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
It is bad idea to have duplicate packages and the preference is to maintain types outside DT. So this package needs to be removed from DT instead of accepting this PR.

**Pull request closed without merging**
